### ISSUE TITLE
CMake?

### ISSUE THREAD

**Issue by cody271** [libui-ng/libui-ng#42]
Instead of using Meson, should we switch _back_ to CMake?

**Comment by rubyFeedback** [comment]
Hmm. Personally I prefer meson, but to me it's not important either way, so however you feel the maintenance load is kept low is fine. I'd retain the meson-build variant or store it somewhere in the event people may want to use it. Some projects offer both meson and cmake. Libui is probably so small that it should not make a huge difference either.

Perhaps what may be helpful/useful, either way, is to list some real tangible benefits of either meson OR cmake. Perhaps they differ in how to work in a cross-platform way.

The reason I like meson more is mostly that I find meson sort of is a spiritual successor to GNU configure. But I am fine either way. I prefer both meson and cmake over GNU configure usually too these days. m4 macros are sooooooooo ugly ...

**Comment by szanni** [comment]
I personally definitely preferred CMake. But to be honest I can't stand any of the options out there, be it CMake, meson, autotools,...
The only thing that really works for me are plain old Makefiles. But that is a nightmare for cross platform development.

I actually have a Makefile in place to run the meson & ninja commands for when I run `make` because I can not for the life of it remember any of the commands or don't want to type in `ninja -C build` when I want to compile. Who thought that is reasonable behavior in the first place?

**Comment by msink** [comment]
Meson can do cross-compilation by adding just  one option `--cross-file cross_file.txt`.
In CMake cross-compilation is far more complex, if possible at all.

**Comment by cody271** [comment]
> Meson can do cross-compilation by adding just one option `--cross-file cross_file.txt`. In CMake cross-compilation is far more complex, if possible at all.
>
We need to add an [ARM build](https://github.com/libui-ng/libui-ng/issues/41), and neither build tool supports this easily on multiple platforms sadly..

CMake: Supports ARM automatically using Apple-specific `clang` flags, only on macOS.
(`CMAKE_OSX_ARCHITECTURES=arm64;x86_64` for fat binary)

Meson: In theory, `--cross-file` would work on all platforms. Except you have to write all the platform-specific steps manually in Meson .py DSL .. not just a change of the config file.

**Comment by msink** [comment]
In meson `cross-file` just describes your cross-toolchain, and no modifications needed in main DSL build file.
See https://mesonbuild.com/Cross-compilation.html
QEMU here will be needed only for running tests, if any.

**Comment by cody271** [comment]
> In meson `cross-file` just describes your cross-toolchain, and no modifications needed in main DSL build file. See https://mesonbuild.com/Cross-compilation.html QEMU here will be needed only for running tests, if any.

Is there a good example of a Meson `cross_file.txt` somewhere that does what we want? A Linux build with both x86 native compile, and cross compile to ARM?

(I think macOS specific ARM build does require a lot of manual steps in the DSL build file. No support for this in Meson.)

**Comment by msink** [comment]
Well, I tested meson cross-compilation in https://github.com/msink/hello-c
But it was years ago and on windows.

What do you need exactly?
Compiling on macOS for Raspberry Pi, or what?

**Comment by cody271** [comment]
> 
> What do you need exactly? Compiling on macOS for Raspberry Pi, or what?
>
Much more simple. Just pre-built binaries in CI for both x86 and ARM, on both Linux and macOS.

**Comment by MikeWang000000** [comment]
> > Meson can do cross-compilation by adding just one option `--cross-file cross_file.txt`. In CMake cross-compilation is far more complex, if possible at all.
> 
> We need to add an [ARM build](https://github.com/libui-ng/libui-ng/issues/41), and neither build tool supports this easily on multiple platforms sadly..
> 
> CMake: Supports ARM automatically using Apple-specific `clang` flags, only on macOS. (`CMAKE_OSX_ARCHITECTURES=arm64;x86_64` for fat binary)
> 
> Meson: In theory, `--cross-file` would work on all platforms. Except you have to write all the platform-specific steps manually in Meson .py DSL .. not just a change of the config file.

It can be achieved by:
```
libui_project_compile_args += ['-arch', 'arm64', '-arch', 'x86_64']
libui_project_link_args += ['-arch', 'arm64', '-arch', 'x86_64']
```
It's not very elegant though.

May be we need a `meson_options.txt`?

**Comment by jcelerier** [comment]
> Meson can do cross-compilation by adding just one option `--cross-file cross_file.txt`. In CMake cross-compilation is far more complex, if possible at all.

how is this different than CMake's, -DCMAKE_TOOLCHAIN_FILE=<blabla.txt> which is exactly the same ? see for instance this project with a wide array of pre-existing cross toolchains for cmake: https://github.com/dockcross/dockcross

**Comment by kojix2** [comment]
If pull request #198 is merged, will Zig replace Meson?

**Comment by szanni** [comment]
> If pull request #198 is merged, will Zig replace Meson?

I do not believe that to be the intention.

For anyone interested in building with cmake, [here my cmake branch](https://github.com/szanni/libui-ng/tree/cmake).
It does work with linux, msvc, mingw and on darwin (you need to disable `-DBUILD_EXAMPLES=off` due to come `libc++` issues).

**Comment by cody271** [comment]
> > If pull request #198 is merged, will Zig replace Meson?
> 
> I do not believe that to be the intention.

Correct, Zig would not be replacing anything. Although Zig might become the recommended cross compiler solution, and *maybe* even the recommended compiler for all Windows development. (The Zig DX is much better than anything else.)

One of the benefits of such a minimalist library is that maintaining multiple build systems should be possible!
